src/net.cpp

```diff
@@ -1107,61 +1107,70 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
-            BOOST_FOREACH(CNode* pnode, vNodesCopy)
-            {
-                if (pnode->fDisconnect)
-                {
-                    // remove from vNodes
-                    vNodes.erase(remove(vNodes.begin(), vNodes.end(), pnode), vNodes.end());
+            // remove from vNodes
+            vNodes.erase(remove(vNodes.begin(), vNodes.end(), pnode), vNodes.end());
 
-                    // release outbound grant (if any)
-                    pnode->grantOutbound.Release();
+            // release outbound grant (if any)
+            pnode->grantOutbound.Release();
 
-                    // close socket and cleanup
-                    pnode->CloseSocketDisconnect();
+            // close socket and cleanup
+            pnode->CloseSocketDisconnect();
 
-                    // hold in disconnected pool until all refs are released
-                    pnode->Release();
-                    vNodesDisconnected.push_back(pnode);
-                }
-            }
+            // hold in disconnected pool until all refs are released
+            pnode->Release();
+            vNodesDisconnected.push_back(pnode);
         }
-        {
-            // Delete disconnected nodes
-            std::list<CNode*> vNodesDisconnectedCopy = vNodesDisconnected;
-            BOOST_FOREACH(CNode* pnode, vNodesDisconnectedCopy)
+    }
+}
+
+void CConnman::DeleteDisconnectedNodes()
+{
+    std::list<CNode*> vNodesDisconnectedCopy = vNodesDisconnected;
+    BOOST_FOREACH(CNode* pnode, vNodesDisconnectedCopy)
+    {
+        // wait until threads are done using it
+        if (pnode->GetRefCount() <= 0) {
+            bool fDelete = false;
             {
-                // wait until threads are done using it
-                if (pnode->GetRefCount() <= 0) {
-                    bool fDelete = false;
-                    {
-                        TRY_LOCK(pnode->cs_inventory, lockInv);
-                        if (lockInv) {
-                            TRY_LOCK(pnode->cs_vSend, lockSend);
-                            if (lockSend) {
-                                fDelete = true;
-                            }
-                        }
-                    }
-                    if (fDelete) {
-                        vNodesDisconnected.remove(pnode);
-                        DeleteNode(pnode);
+                TRY_LOCK(pnode->cs_inventory, lockInv);
+                if (lockInv) {
+                    TRY_LOCK(pnode->cs_vSend, lockSend);
+                    if (lockSend) {
+                        fDelete = true;
                     }
                 }
             }
+            if (fDelete) {
+                vNodesDisconnected.remove(pnode);
+                DeleteNode(pnode);
+            }
         }
+    }
+}
+
+void CConnman::ThreadSocketHandler()
+{
+    unsigned int nPrevNodeCount = 0;
+    const unsigned int nUnevictableConnections = std::max(0, std::max(MAX_OUTBOUND_CONNECTIONS, MAX_ADDNODE_CONNECTIONS) - PROTECTED_INBOUND_PEERS);
```

Comment:
Why is this `max_addnode - protected_inbound` and not `max_addnode + protected_inbound`?

```
MAX_OUTBOUND_CONNECTIONS = 8
MAX_ADDNODE_CONNECTIONS = 8
PROTECTED_INBOUND_PEERS = 20

std::max(8, 8) = 8
8 - 20 = -12
std::max(0, -12) = 0
```

Or am I missing something?

Reply:
That also seems correct. I'll push a fix.

In my experience with this PR, if I were to design the right way to accomplish this, I wouldn't start with the code as inherited from Bitcoin. The constants and primitives aren't quite right to do this as cleanly as I'd like. I think we'll end up with something usable though, and there are a couple of further improvements we can make now that we see their limitations.

Reply:
> if I were to design the right way to accomplish this, I wouldn't start with the code as inherited from Bitcoin

I feel you. The eternal balance between compatibility with upstream and higher quality code. Perhaps we can do this right on a 1.21 port, assuming that we have a little bit more discussion on #2273 and create a guideline for this.

For 1.14... I still expect to have a moment of starting to wind down in favor of 1.21 later this year.